scaletest/notifications/run.go

```diff
@@ -0,0 +1,269 @@
+		go func() {
+			r.cfg.OwnerWatchBarrier.Wait()
+			close(done)
+		}()
+
+		select {
+		case <-done:
+			logger.Info(ctx, "owner users complete, closing connection")
+		case <-ctx.Done():
+			logger.Info(ctx, "context canceled, closing connection")
+		}
+		return nil
+	}
+
+	logger.Info(ctx, "waiting for notifications", slog.F("timeout", r.cfg.NotificationTimeout))
+
+	watchCtx, cancel := context.WithTimeout(ctx, r.cfg.NotificationTimeout)
+	defer cancel()
+
+	if err := r.watchNotifications(watchCtx, conn, newUser, logger, r.cfg.ExpectedNotifications); err != nil {
+		return xerrors.Errorf("notification watch failed: %w", err)
+	}
+
+	reachedOwnerWatchBarrier = true
+	r.cfg.OwnerWatchBarrier.Done()
+
+	return nil
+}
+
+func (r *Runner) Cleanup(ctx context.Context, id string, logs io.Writer) error {
+	if r.createUserRunner != nil {
+		_, _ = fmt.Fprintln(logs, "Cleaning up user...")
+		if err := r.createUserRunner.Cleanup(ctx, id, logs); err != nil {
+			return xerrors.Errorf("cleanup user: %w", err)
+		}
+	}
+
+	return nil
+}
+
+const NotificationDeliveryLatencyMetric = "notification_delivery_latency_seconds"
+
+func (r *Runner) GetMetrics() map[string]any {
+	metrics := map[string]any{}
+
+	for templateID, latency := range r.notificationLatencies {
+		if latency > 0 {
+			metricKey := fmt.Sprintf("%s_%s", NotificationDeliveryLatencyMetric, templateID.String())
+			metrics[metricKey] = latency.Seconds()
+		}
+	}
+
+	return metrics
+}
+
+func (r *Runner) dialNotificationWebsocket(ctx context.Context, client *codersdk.Client, user codersdk.User, logger slog.Logger) (*websocket.Conn, error) {
+	u, err := client.URL.Parse("/api/v2/notifications/inbox/watch")
+	if err != nil {
+		logger.Error(ctx, "parse notification URL", slog.Error(err))
+		r.cfg.Metrics.AddError(user.Username, "parse_url")
+		return nil, xerrors.Errorf("parse notification URL: %w", err)
+	}
+
+	conn, resp, err := websocket.Dial(ctx, u.String(), &websocket.DialOptions{
+		HTTPHeader: http.Header{
+			"Coder-Session-Token": []string{client.SessionToken()},
+		},
+	})
+	if err != nil {
+		if resp != nil {
+			defer resp.Body.Close()
+			if resp.StatusCode != http.StatusSwitchingProtocols {
+				err = codersdk.ReadBodyAsError(resp)
+			}
+		}
+		logger.Error(ctx, "dial notification websocket", slog.Error(err))
+		r.cfg.Metrics.AddError(user.Username, "dial")
+		return nil, xerrors.Errorf("dial notification websocket: %w", err)
+	}
+
+	return conn, nil
+}
+
+// watchNotifications reads notifications from the websocket and returns error or nil
+// once all expected notifications are received.
+func (r *Runner) watchNotifications(ctx context.Context, conn *websocket.Conn, user codersdk.User, logger slog.Logger, expectedNotifications map[uuid.UUID]chan time.Time) error {
+	logger.Info(ctx, "waiting for notifications",
+		slog.F("username", user.Username),
+		slog.F("expected_count", len(expectedNotifications)))
+
+	receivedNotifications := make(map[uuid.UUID]bool)
```

Comment:
nit:
```suggestion
	receivedNotifications := make(map[uuid.UUID]struct{})
```
`struct{}` is zero-size, bool is a byte